### ISSUE TITLE
Unable to compile: Duplicate interface definition for class 'FLAnimatedImageView'

### ISSUE THREAD

**Issue by DerJacques** [DylanVann/react-native-fast-image#46]
Hi there!

I'm having problems compiling my app after adding react-native-fast-image.

After running `react-native link react-native-fast-image`, I receive a bunch of `duplicate interface definition for class 'FLAnimatedImageView'` and `property has a previous declaration` errors.

Here's a subset of the errors I receive:

```
In file included from /Users/manuelthomsen/Coding/experiments/expo/woodn/node_modules/react-native-fast-image/ios/Vendor/SDWebImage/SDWebImage/FLAnimatedImage/FLAnimatedImageView+WebCache.m:16:
In file included from /Users/manuelthomsen/Coding/experiments/expo/woodn/node_modules/react-native-fast-image/ios/Vendor/SDWebImage/Vendors/FLAnimatedImage/FLAnimatedImage/FLAnimatedImage.h:13:
/Users/manuelthomsen/Coding/experiments/expo/woodn/node_modules/react-native-fast-image/ios/Vendor/SDWebImage/Vendors/FLAnimatedImage/FLAnimatedImage/FLAnimatedImageView.h:22:1: error: duplicate interface definition for class 'FLAnimatedImageView'
@interface FLAnimatedImageView : UIImageView
^
In module 'FLAnimatedImage' imported from /Users/manuelthomsen/Coding/experiments/expo/test/node_modules/react-native-fast-image/ios/Vendor/SDWebImage/SDWebImage/FLAnimatedImage/FLAnimatedImageView+WebCache.h:14:
/Users/manuelthomsen/Library/Developer/Xcode/DerivedData/test-ftzdxdyafvkexabeahltbammdnfq/Build/Products/Debug-iphonesimulator/FLAnimatedImage.framework/Headers/FLAnimatedImageView.h:22:12: note: previous definition is here
@interface FLAnimatedImageView : UIImageView
           ^
In file included from /Users/manuelthomsen/Coding/experiments/expo/test/node_modules/react-native-fast-image/ios/Vendor/SDWebImage/SDWebImage/FLAnimatedImage/FLAnimatedImageView+WebCache.m:16:
In file included from /Users/manuelthomsen/Coding/experiments/expo/test/node_modules/react-native-fast-image/ios/Vendor/SDWebImage/Vendors/FLAnimatedImage/FLAnimatedImage/FLAnimatedImage.h:13:
/Users/manuelthomsen/Coding/experiments/expo/test/node_modules/react-native-fast-image/ios/Vendor/SDWebImage/Vendors/FLAnimatedImage/FLAnimatedImage/FLAnimatedImageView.h:26:48: error: property has a previous declaration
@property (nonatomic, strong) FLAnimatedImage *animatedImage;
```

Any ideas on how to solve this?
This is a basic expo app that has just been detached.

Tested on RN v. 46 and 47.

Thank you!

**Comment by fredzhang** [comment]
same problem here. please help

**Comment by mortezaalizadeh** [comment]
Same problem here :)

**Comment by denny64** [comment]
+1

**Comment by jacksontbryan** [comment]
Blowing away the contents of my ios/build folder cleared this for me.  The error only happened from react-native run-ios and wasn't reproducible from xcode directly.

**Comment by rjacobskind** [comment]
@DylanVann I'm having this issue as well. Any thoughts on how to solve the compile error? I'm getting a build fail both from react-native run-ios and from xcode

**Comment by srameshr** [comment]
@jacksontbryan Did that. It did not solve the problem for me. It occurs when building for iPhone 5s and iOS version 9.0

**Comment by MrLoh** [comment]
I'm also encountering this issue. No idea how to fix it, or where it suddenly came form.

**Comment by ruipaulo** [comment]
Product-> Clean and then 'rm -rf ~/Library/Developer/Xcode/DerivedData/*' did the trick for me.

**Comment by i6mi6** [comment]
Cmd + Shift + K did it for me

**Comment by jgbaEmento** [comment]
same problem here ... cleaning and deleting derived data does not work here

**Comment by kasran15** [comment]
Product -> Clean did the trick. Good suggestion!

**Comment by rouge3351** [comment]
I’ve just delete project/ios/build folder and command react-native run-ios  and it works fine!

**Comment by bonlemuel** [comment]
@rouge3351 Thanks for this. Saved me lots of time figuring it out. :)

**Comment by luco** [comment]
None of proposed fixes works for me.

**Comment by MrLoh** [comment]
@luco did you really try all three things? 
- `rm -rf ~/Library/Developer/Xcode/DerivedData/*`
- `rm -rf ios/build`
- in xcode <kbd>Cmd</kbd> + <kbd>Shift</kbd> + <kbd>K</kbd>

Did you try building from xcode and command line

@DylanVann is there any chance you could look into fixing this in the code. This is sucha a terrible annoyance for everyone using this library and it happens regularly. I have no idea what has to be setup in xcode to make this disappear, but there must be something, as I haven't gotten these errors from any other library.

**Comment by luco** [comment]
@MrLoh Apparently doing this steps the error went away. I'll be monitoring this. Thanks a lot!

**Comment by hoscarcito** [comment]
This is really annoying indeed.
I'm subscribing for news!

**Comment by broskoski** [comment]
Same problem here. Running Project -> Clean and wiping the derived data folder did not work for me. If it helps at all, my project is set up with CocoaPods (though I manually linked the library in the build phases step).

**Comment by ruipaulo** [comment]
@broskoski Did you do all the above suggested steps in @MrLoh 's last comment?

**Comment by broskoski** [comment]
@ruipaulo thanks, yes I did.

**Comment by danmatthews** [comment]
Also had this problem and couldn't solve it by performing all of the above steps.

I'm on a detached expo project, which means that i don't have an `ios/build` folder, so i couldn't do that one.

Had to uninstall as it was preventing me from building entirely.

**Comment by DylanVann** [comment]
- node_modules
- react-native packager cache
- xcode clean
- xcode derived data
- react native `ios/build`

There are a lot of places things can get cached and mess things up when using React Native. Try clearing all those and I'm certain you won't have any more issues.

There's not really anything I could do to this library that would prevent these issues. I have dealt with these issues using many other libraries.

IMO React Native should have a clean command that handles all of these cases.

**Comment by ruipaulo** [comment]
@DylanVann I know, that it is probably super hard to find the bug that is causing this. On the other hand, it is happening consistently once or twice per week in our project. This is only and always caused by fast-image (we have also many other native libraries in the project). I know that there is probably no easy fix for this, but I think this issue should remain open as long as it is not fixed. Maybe at some point somebody might come up with a solution. This is a relevant bug, it is costing us at least an hour dev time per week and many users coming new to your library might spent hours trying to fix find a solution. This shouldn't be perceived as nagging, we love your library, but this is a relevant bug.

**Comment by foggy1** [comment]
This happens in completely fresh build environments, e.g. when using continuous integration tools. It actually makes CI completely untenable with respect to iOS builds.

**Comment by DannyAlb** [comment]
I do not recommend doing this, but our extremely unacceptable and voodoo workaround was that after trying to build normally and getting above errors, to build separately, and add the library statically to the project (the libFastImage.a).

Unacceptable, voodoo, but the only way we got it to works :)

**Comment by brenwell** [comment]
I am wondering if the issue is caused because there is actually a duplicate?

Looks like its included twice.

> I'm no xcode expert though

<img width="394" alt="screen shot 2018-04-06 at 09 05 13" src="https://user-images.githubusercontent.com/802618/38407815-45d3bb0a-397b-11e8-9b23-1f3f8fa20267.png">

**Comment by hoscarcito** [comment]
As @foggy1 mention, this happens on fresh build environments, making it useless with CI tools. 
@brenwell I saw the same but I don't understand how it works iOS nor why it works after cleaning the project.
Any light over here?

**Comment by tomauty** [comment]
@brenwell They are the same file. Not sure why it's included twice but if that could be resolved, all users here could finally check `Parallelize Builds` again...

**Comment by oferRounds** [comment]
+1

**Comment by DylanVann** [comment]
I'm looking into it a bit more. I have run into this.

I cannot replicate the issue on a new project though.

**Comment by DylanVann** [comment]
I made some changes to the `FastImage` project that might fix this.

- Set deprecated ALWAYS_SEARCH_USER_PATHS to NO. (e7ba4a7f789d883f4dbbe526612e70a2501d7be5)
- Fix `FLAnimatedImage` header search path. (883dc0664dfd6ca26a1b8bece161abd3b9184cf1)
- Remove `FLAnimatedImage` from `FastImage` project since it's already included in `SDWebImage`. Installation remains the same as before when using CocoaPods. (a2d9fe2c71693721fec56e9cfe258a373a651b71)

The changes are published as `v4.0.4`. Please let me know wether or not you're still experiencing this issue after upgrading.

**Comment by brenwell** [comment]
Awesome. Will check thanks

**Comment by VicFrolov** [comment]
Getting same issue on version 4.0.4

Removed build folder, removed node modules, cleared watchman, installed node modules, npm reset cache, deleted all derived data. cleared xcode cache. Not sure what else to do :/

**Comment by brenwell** [comment]
I haven't seen the issue since I upgraded to 4.0.4 that doesn't mean it won't appear though

For anyone interested here is my **clean all the things** script

```sh
# react-native
    watchman watch-del-all
    rm -rf node_modules && npm install
    rm -rf $TMPDIR/haste-map-react-native-packager-*
    rm -rf $TMPDIR/react-native-packager-cache-*
    rm -rf $TMPDIR/metro-bundler-cache-*

# xcode
    xcodebuild -project ios/vs2.xcodeproj -configuration Release clean
    xcodebuild -project ios/vs2.xcodeproj -configuration Debug clean
    rm -rf ~/Library/Developer/Xcode/DerivedData
    rm -f /ios/build

 # cocoapods
    pod deintegrate --project-directory=./ios/
    rm -rf ./ios/Pods
    pod install --project-directory=./ios/
```

**Comment by DylanVann** [comment]
There were some issues with `v4.0.4`, they should be resolved in the latest version (`v4.0.6`). See: https://github.com/DylanVann/react-native-fast-image/issues/189

**Comment by Ingibjorg** [comment]
Thanks @DylanVann . This solved the issue for me 🙏

**Comment by schlaegerz** [comment]
I am on 4.0.8 and still getting this issue

**Comment by DylanVann** [comment]
You will probably still have this issue immediately after you upgrade to `v4.0.8`.

Do a clean build (xcode clean / `rm -rf ios/build`). I haven't seen this issue again since doing that.

**Comment by luco** [comment]
OK, I got this issue again and solved by doing a xcode clean and removing `ios/build` folder.

**Comment by duro** [comment]
@DylanVann we are seeing this on latest version of react-native-fast-image with React 0.57. Not sure if 0.57 has anything to do with it. The biggest problem is that we use CI to build our releases, which always start from a clean state. We are consistently getting the issue with every build.

One thing we have noticed is that another one of our dependencies, the Braze/Appboy SDK, also depends on SDWebImage.

Has any further light been shed on this issue? We are going crazy trying to make any progress on resolving this.

**Comment by duro** [comment]
@dmkmedina: Pinging @theonetheycallneo as he was the one on our team who got it sorted out. I'll be honest, it wasn't straightforward.

**Comment by onurzorluer** [comment]
Xcode 10 has a new build system. I had same problem in Xcode 10. I found that switching to the Legacy Build Systems solved the problem.
To switch to the legacy build system do the following:

- [ ] File -> Project/Workspace Settings...
- [ ] Under Build System choose: Legacy Build System

**Comment by theonetheycallneo** [comment]
@dmkmedina Our solution wasn't ideal, but we forked the react-native-fast-image and stripped out their SDWebImage.framework, then had their iOS project inherit from our parent umbrella iOS project:
https://github.com/FreebirdRides/react-native-fast-image/commits/master

We then manually linked our parent iOS project to SDWebImage with Appboy's latest bundled release of SDWebImage.

Basically, we been manually linking frameworks on our umbrella, then we fork any conflicting libs and have them correctly inherit from the parent.

Hope that helps shed some light.
